Remove usage of Session from adapter

dariorussi opened the pull request:
## Description 

This is the first step in removing `Session` from the execution layer/adapter.
It actually gets rid of `Session` entirely, however more clean up is needed.
I am looking for feedback if anybody has strong feelings and have thought about how to refactor the VM component. What I mean is that we have few "objects" that contribute to Move execution: `MoveVM`, `NativeContextExtensions`, `LinkageView`, `ExecutionState`, and they are currently all splashed as single fields in `ExecutionContext`. I feel that they may benefit from being abstracted as a single component/struct with a cohesive API (that again, maybe not).
I will work in that direction but I would welcome ideas if anybody has thought about this already

## Test Plan 

Existing tests, this is not a new feature and it is not intended to introduce any new behavior

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -874,20 +866,24 @@ fn init_modules<Mode: ExecutionMode>(
+                "init should not have return values"
+            ),
+            Err(err) => {
+                context.linkage_view.unpublish_package(modules);
```
Woah...

tnowacki replied:
Is this new? Or something the session automatically did?

dariorussi replied:
yeah that is the part of session we were using and I just shoveled it there for now.
Not clear it will stay there.
But basically module init needed that code. I plan to review the ordeal of module initializers so we'll see...
but yeah that is the new non-sense.
Maybe we should fined a better way to do that. Thoughts? Feelings?

dariorussi replied:
in fact come to think of it this is a question for @amnn on how he may want to get this done.
I added the list of modules here just because it was simple and easy. But we have the list of published packages in context, maybe we should reconcile?

dariorussi replied:
Moved the `DataStore` implementation close to the `ExecutionContext` and the new packages tracked there

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/linkage_view.rs:
```diff
@@ -279,6 +276,43 @@ impl<'state> LinkageResolver for LinkageView<'state> {
     }
+
+    pub(crate) fn unpublish_package(&mut self, modules: &[CompiledModule]) {
+        let module_ids: Vec<ModuleId> = modules.iter().map(|m| m.self_id()).collect();
+        for module_id in module_ids {
```
Why do you need to collect here?
Could this just be 
```suggestion
        for module_id in modules.iter().map(|m| m.self_id()) {
```

dariorussi replied:
I guess so.... lol, we'll change

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/linkage_view.rs:
```diff
@@ -279,6 +276,43 @@ impl<'state> LinkageResolver for LinkageView<'state> {
     }
+
+    pub(crate) fn unpublish_package(&mut self, modules: &[CompiledModule]) {
+        let module_ids: Vec<ModuleId> = modules.iter().map(|m| m.self_id()).collect();
+        for module_id in module_ids {
+            self.published_modules.remove(&module_id);
+        }
+    }
```
Or just
```suggestion
        for module in modules {
            self.published_modules.remove(&module.self_id());
        }
    }
```

tnowacki reviewed (commented):
Small nit, I don't think a lot of Rust APIs do the `get_` prefix, so `get_vm_runtime()` would just be `vm_runtime()` and similarly for `get_object_runtime()`

dariorussi commented on external-crates/move/move-vm/runtime/src/runtime.rs:
```diff
@@ -35,16 +35,19 @@ use move_vm_types::{
 impl VMRuntime {
     pub(crate) fn new(
-        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        natives: impl IntoIterator<Item=(AccountAddress, Identifier, Identifier, NativeFunction)>,
```
possible that some linting did that.... I'll fix it back and see

dariorussi commented:
addressed feedback

amnn commented on sui-execution/latest/sui-adapter/src/programmable_transactions/linkage_view.rs:
```diff
@@ -40,6 +48,8 @@ pub struct LinkageView<'state> {
     /// set, then we will not try to load its type origin table when setting it as a context (again).
     past_contexts: RefCell<HashSet<ObjectID>>,
+    /// Modules published for the session using this LinkageView.
+    published_modules: BTreeMap<ModuleId, Vec<u8>>,
```
Chatted offline about consolidating this in `ExecutionContext`.

dariorussi replied:
Created a `SuiDataStore` that works with `ExecutionContext` to use new packages

amnn reviewed (commented):
Looks good pending the stuff we've already talked about!  Looking squeaky clean.

dariorussi commented on crates/sui-types/src/object.rs:
```diff
@@ -679,6 +679,11 @@ impl Object {
         }
     }
 
+    // Note: this will panic if `modules` is empty
```
I'll double check as this is something that was there already I did not want to change

dariorussi replied:
I removed the comment as I cannot spot an obvious `panic` around. 
Though it is a bit involving as that flows in a couple of places. They all seem to be properly protected though

amnn replied:
Yeah, this is no longer true, after hardening -- it's `MovePackage::new_package` and/or `MovePackage::new_upgraded_package` that this used to apply to, but these now return a `Result`.

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1323,4 +1323,80 @@ unsafe fn create_written_object<'vm, 'state>(
+        }
+    }
+
+    fn publish_module(&mut self, _module_id: &ModuleId, _blob: Vec<u8>) -> VMResult<()> {
```
no we cannot panic here because the VM/runtime calls this. At a later point in time we can (and most likely will) change the traits and callback used for the `DataStore`, but the way things are now that is a call we just have to do nothing, or we hook up instead of the work we do in publishing

dariorussi commented:
more feedback, moved `DataStore` into `ExecutionContext` (removed from `LinkageView`).
We'll see how it goes....

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -245,43 +239,42 @@ impl<'vm, 'state, 'a> ExecutionContext<'vm, 'state, 'a> {
     pub fn load_type(&mut self, type_tag: &TypeTag) -> VMResult<Type> {
-        load_type(&mut self.session, type_tag)
+        load_type(self.vm, &mut self.linkage_view, type_tag)
     }
```
removed though I left `load_type` as a simple helper

dariorussi commented:
addressed all feedback, commented code to clarify things.
This is, in principle, good to go (though I will sync a full node over the weekend to double check), but I am going to wait for @tnowacki to get back anyway before committing even if accepted.
As mentioned this is the first step and I'll keep building on this with other commits so no big deal if it does not go in right away.
We really need to clean up more and structure the code better but it is no an easy feat and it requires some thinking and incremental work.
All things considered this removes `Session` and could be a decent first step

dariorussi commented:
there are some white space changes showed in the PR that I am not sure I know how to remove. So I am letting them be

dariorussi commented on crates/sui-types/src/move_package.rs:
```diff
@@ -382,6 +383,18 @@ impl MovePackage {
         )
     }
 
+    // Retrieve the module with `ModuleId` in the goven pacakge.
+    // Check is the address of the module is the same of the package
+    // and return `None` if that is not the case.
+    // All modules in a package share the address with the package.
```
@amnn is this comment even true?

amnn replied:
Nope. Is this ported from elsewhere?  It's generally a good idea to explicitly indicate whether `module_id` here is a `runtime_id` or a `storage_id` and document this in the doc comment as well, then it would be clearer whether this was implemented correctly.  If it's a `storage_id`, then the implementation is correct.  If it's a `runtime_id`, it's not, and the correct implementation is to fetch the module from the map first and check whether its self address matches the `ModuleId`'s address.

dariorussi replied:
I wrote that comment :) and the function, so I may very well have written it wrong, let me go and check the callers.
Thanks!

dariorussi replied:
@amnn see if this make sense to you and changing the comment.
The only call to this is from the loader here https://github.com/MystenLabs/sui/blob/main/external-crates/move/move-vm/runtime/src/loader.rs#L1073-L1081 where it explicitly uses the `storage_id`, so I guess we are good. 
Thoughts?

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -521,22 +499,26 @@ mod checked {
-            Ok(())
+        }
+
+        /// Return the last pacakge pushed in `write_package`.
+        /// This function should be used in block of codes that push a pacakge, verify
+        /// it, run the init and in case of error will remove the package.
+        /// The package has to be pushed for the init to run correctly.
```
you can check the caller to see what this means. 
Not pretty and happy to get feedback if someone feels strong about it, but I think it's a fine step for now.
Once we plug in the loader in a more structured and reasonable way this may all go away

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1345,4 +1344,118 @@ mod checked {
         )
     }
+
+    // Implementation of the `DataStore` trait for the Move VM.
```
for better or worse this is it

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -493,27 +494,29 @@ mod checked {
-        context.set_linkage(package)?;
+        let package = context.new_package(&modules, &dependencies)?;
+
+        // Here we optimistacally push the package that is being published/upgraded
+        // and if there is an error of any kind (verification or module init) we
+        // remove it.
+        // The call to `pop_last_package` later is fine because we cannot re-enter and
+        // the last package we pushed is the one we are verifying and running the init from
```
this is the part I mentioned in the comment about in `pop_last_package`.
Let me know if you are fine with it, or you want to throw a tantrum...

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/linkage_view.rs:
```diff
@@ -281,6 +267,36 @@ impl<'state> LinkageResolver for LinkageView<'state> {
+    }
+}
+
+impl<'state> LinkageResolver for LinkageView<'state> {
```
I think this is not needed any longer strictly speaking and it was needed when `DataStore` was implemented here.
However I have the feeling it will help in the future to have the implementation of those function in the `impl` for `LinkageView` so I'm leaving the change in.
Let me know if you are bothered too much by this

amnn reviewed (commented):
This change looks good to go from my perspective, but will wait for at least @tnowacki to take a look as well.

wlmyng commented on crates/sui-types/src/move_package.rs:
```diff
@@ -382,6 +383,19 @@ impl MovePackage {
+    // and return `None` if that is not the case.
+    // All modules in a package share the address with the package.
+    pub fn get_module(&self, storage_id: &ModuleId) -> Option<&Vec<u8>> {
+        if self.id != ObjectID::from(*storage_id.address()) {
+            None
+        } else {
+            self.module_map.get(&storage_id.name().to_string())
+        }
```
🤯

dariorussi replied:
I may drop the suggestion, though thanks, because it's not reading that much better for me :)

dariorussi commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1346,4 +1345,118 @@ mod checked {
+            if let Some(bytes) = self.get_module(module_id) {
+                return Ok(bytes.clone());
+            }
+            match self.linkage_view.get_module(module_id) {
```
I am not too sure to be honest and I was thinking to tackle that next as a iterative cleanup of the traits used.
I believe `TypeLayoutResolver` gets in the way for few things but overall I am not clear that `LinkageView` needs much

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1084,26 +1064,32 @@ mod checked {
         let tag: StructTag = type_.into();
-        let type_ = load_type(session, &TypeTag::Struct(Box::new(tag)))
-            .map_err(|e| crate::error::convert_vm_error(e, vm, session.get_resolver()))?;
+        let type_ = load_type(
```
would be nice if we had `load_struct_type` or something exposed, just to avoid this allocation (and in some cases a cloning). But it is fine for now :)

dariorussi replied:
man it took me 20 minutes to understand what we were talking about, finally got it - lol...
take a look and see if it makes sense

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1346,4 +1345,118 @@ mod checked {
+        }
+
+        fn get_module(&self, module_id: &ModuleId) -> Option<&Vec<u8>> {
+            for package in self.new_packages {
```
What's our cap on new packages?

dariorussi replied:
it' a good question let me think about this a bit...

dariorussi replied:
made some changes, see if they make sense to you

dariorussi commented:
some feedback still checking some final things on new packages...

dariorussi commented:
addressed feedback and made some change in the way we handle `new_packages`

amnn commented on sui-execution/latest/sui-adapter/Cargo.toml:
```diff
@@ -11,6 +11,7 @@ edition = "2021"
 anyhow = { workspace = true, features = ["backtrace"] }
 bcs.workspace = true
 leb128.workspace = true
+linked-hash-map.workspace = true
```
Actually, ignore me, because I see that we are already using `LinkedHashMap` in `sui-types` for execution-related stuff.

tnowacki replied:
I personally think IndexMap has a slightly better API, though its all about the same :)

dariorussi replied:
I'll look at `IndexMap` though as pointed out would still increase dependencies. 
I am fine with the added dependency fwiw, not sure about others

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -68,19 +81,20 @@ mod checked {
         additional_transfers: Vec<(/* new owner */ SuiAddress, ObjectValue)>,
         /// Newly published packages
-        new_packages: Vec<Object>,
+        new_packages: LinkedHashMap<ObjectID, MovePackage>,
```
I think you would get away just fine with a `BTreeMap`, since it all get's thrown into a `BTreeMap` at the end. But it feels nice to preserve the order if we can I suppose

amnn replied:
The reason we need the order here is so that we can "unpublish".

tnowacki replied:
Why do you need the order for that?

amnn replied:
Although yes, you can do this just by remembering the package ID that you want to unpublish and passing that in!

tnowacki commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -522,22 +511,26 @@ mod checked {
-            self.new_packages.push(package);
-            Ok(())
+        pub fn write_package(&mut self, package: MovePackage) {
+            self.new_packages.insert(package.id(), package);
```
+1 because why not?

dariorussi replied:
we could, and you are not being paranoid, it is in fact a question I asked myself too and discussed with Todd.
I should have talked to you as well about it come to think of it.
In any case it did not seem possible all things considered which could be a reason to add the invariant violation?

dariorussi commented:
fwiw I am moving this to a new version of the execution and add some changes to how we charge for publishing/upgrading.
Adding those changes would require a protocol config bump and we thought it may be fine to bundle that with this change and have a new version of execution. In a sense it would be "safer" somewhat.
cc @tnowacki

dariorussi commented:
I am going to abandon this in favor of https://github.com/MystenLabs/sui/pull/13478

dariorussi commented:
this is going away in favor of #13504
